lib/datasource/pypi/index.ts

```diff
@@ -180,21 +203,26 @@ async function getSimpleDependency(
     logger.trace({ dependency: packageName }, 'pip package not found');
     return null;
   }
-  const root: HTMLElement = parse(dep.replace(/<\/?pre>/, '')) as any;
+  const root: HTMLElement = parse(cleanSimpleHtml(dep)) as any;
   const links = root.querySelectorAll('a');
-  const releases = new Map<string, Release>();
```

Comment:
I think this Map was serving to also de-duplicate. If you shift to an array approach, you will need to add a de-duplication step manually to get equivalent results. However I'm not sure why we had concerns with duplication in the first place.

Reply:
The deduplication now happens in this function https://github.com/renovatebot/renovate/pull/6966/files#diff-de5edfc564f5473b7f2a5a8c48e82629R58-R80, which also takes care of the deduplication when getting dependency from the JSON endpoint. 

I believe the deduplication is needed because a version could have both sdist (tar.gz) and wheel published.

I've updated the test to validate deduplication. The simple endpoint [fixture](https://github.com/renovatebot/renovate/pull/6966/files#diff-531b485f5ecf91b58397404a8b03c674R13-R14) contains both sdist and wheel releases for 0.2.2 and the [response](https://github.com/renovatebot/renovate/pull/6966/files#diff-6d852fa0319be1d392d9f04edba65ea9R101-R103) contains only one entry for 0.0.2

Reply:
`getCompatibleReleases` is intended for comparing compatibility and not de-duplicating. It would be very easy to accidentally regress this in future so I'd prefer it's done explicitly.

Reply:
Oh `getCompatibleReleases` is a function I added in this PR which pulls together the shared logic between `getDependency` and `getSimpleDependency` (now that simple also supports compatibility check) to do the compatibility check by invoking `compatibleVersions`, deduplicate versions and transform dependencies from pypi specific JSON to a more generic `Release` object. Maybe this function can be named better? E.g., `deduplicateAndTransformReleases`?

Reply:
Explicit and maintainable is always better than succinct and clever, so therefore please split it into two steps - compatibility and de-duplication. It will make it much easier for the next person who needs to understand and change the code.

Reply:
👍 Updated to keep compatibility and de-duplication separate.